### PR TITLE
Endpoint to get Builds without attributes.

### PR DESCRIPTION
**Pull request by matejonnet**
### Checklist:

* [ ] Have you added a note in the [CHANGELOG wiki](https://github.com/project-ncl/pnc/wiki/Changelog) for your change if user-facing?
* [x] Have you added unit tests for your change?

### CODE REVIEW

File: facade/src/main/java/org/jboss/pnc/facade/providers/BuildProviderImpl.java [hunk 1]
```diff
@@ -599,6 +605,39 @@ public void setDependentArtifacts(String id, List<String> artifactIds) {
         repository.save(buildRecord);
     }
 
+    public Page<Build> getByAttribute(BuildPageInfo buildPageInfo, Map<String, String> attributeConstraints) {
+        Set<Predicate<BuildRecord>> predicates = new HashSet<>();
+        for (Map.Entry<String, String> entry : attributeConstraints.entrySet()) {
+            String key = entry.getKey();
+            String value = entry.getValue();
+            if (key.startsWith("!")) {
+                withoutAttribute(key);
```
Comment (janinko) [inline facade/src/main/java/org/jboss/pnc/facade/providers/BuildProviderImpl.java]: Should you skip the first character `!`?

File: facade/src/main/java/org/jboss/pnc/facade/providers/BuildProviderImpl.java [hunk 2]
```diff
@@ -599,6 +605,39 @@ public void setDependentArtifacts(String id, List<String> artifactIds) {
...
+            } else {
+                withAttribute(key, value);
+            }
+        }
+
+        Predicate<BuildRecord>[] predicatesArray = predicates.toArray(new Predicate[predicates.size()]);
+
+        PageInfo pageInfo = toPageInfo(buildPageInfo);
+        SortInfo sortInfo = rsqlPredicateProducer.getSortInfo(type, buildPageInfo.getSort());
+        List<BuildRecord> resultList = ((BuildRecordRepository) BuildProviderImpl.this.repository)
```
Comment (janinko) [inline facade/src/main/java/org/jboss/pnc/facade/providers/BuildProviderImpl.java]: You are ignoring the RSQL query. I think it should be also used.

File: facade/src/main/java/org/jboss/pnc/facade/providers/BuildProviderImpl.java [hunk 3]
```diff
@@ -599,6 +605,39 @@ public void setDependentArtifacts(String id, List<String> artifactIds) {
...
+                .queryWithPredicatesUsingCursor(pageInfo, sortInfo, predicatesArray);
+
+        int hits = repository.count(predicatesArray);
+
+        return new Page<>(
+                buildPageInfo.getPageIndex(),
+                buildPageInfo.getPageSize(),
+                (int) Math.ceil((double) hits / buildPageInfo.getPageSize()),
```
Comment (janinko) [inline facade/src/main/java/org/jboss/pnc/facade/providers/BuildProviderImpl.java]: There should be new construcotr that does the page count computation, so you don't need to include this line.

File: rest-api/src/main/java/org/jboss/pnc/rest/api/endpoints/BuildEndpoint.java [hunk 4]
```diff
@@ -91,6 +90,10 @@
     static final String BUILD_STATUS = "Status of the build";
```
Comment (janinko) [inline rest-api/src/main/java/org/jboss/pnc/rest/api/endpoints/BuildEndpoint.java]: Could you add to the `addAttribute` method mention what are the restrictions on the key format?

File: model/src/main/java/org/jboss/pnc/model/BuildRecordAttribute.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+@IdClass(BuildRecordAttribute.AttributeId.class)
+public class BuildRecordAttribute implements Serializable {
+
+    @Id
+    Integer buildRecordId;
+
+    @Id
+    String key;
+
+    String value;
```
Comment (janinko) [inline model/src/main/java/org/jboss/pnc/model/BuildRecordAttribute.java]: private

File: model/src/main/java/org/jboss/pnc/model/BuildRecord.java [hunk 6]
```diff
@@ -673,24 +669,51 @@ public String toString() {
...
+        BuildRecordAttribute attribute = new BuildRecordAttribute(this.id, key, value);
+        if (old.isPresent()) {
+            attributes.remove(old.get());
+        }
+        attributes.add(attribute);
+        if (old.isPresent()) {
+            return old.get().getValue();
+        } else {
+            return null;
+        }
```
[the remark this answers is not in the document]
Comment (matejonnet) [reply model/src/main/java/org/jboss/pnc/model/BuildRecord.java]: It have to return the value of the optional object.

File: model/src/main/java/org/jboss/pnc/model/BuildRecord.java [hunk 7]
```diff
@@ -673,24 +669,51 @@ public String toString() {
...
+    }
+
+    public Optional<BuildRecordAttribute> getAttributeEntity(String key) {
+        return attributes.stream().filter(a -> a.key.equals(key)).findAny();
     }
 
     public String putAttribute(String key, String value) {
-        return attributes.put(key, value);
+        if (this.id == null) {
+            throw new PersistenceException("Build record id must be set before adding the attributes.");
```
[the remark this answers is not in the document]
Comment (matejonnet) [reply model/src/main/java/org/jboss/pnc/model/BuildRecord.java]: I did try with PrePersist to set the id but didn't manage to make it work.

File: model/src/main/java/org/jboss/pnc/model/BuildRecord.java [hunk 8]
```diff
@@ -673,24 +669,51 @@ public String toString() {
...
     }
 
     public String getAttribute(String key) {
-        return attributes.get(key);
+        Optional<BuildRecordAttribute> attribute = getAttributeEntity(key);
+        if (attribute.isPresent()) {
+            return attribute.get().getValue();
+        } else {
+            return null;
+        }
```
[the remark this answers is not in the document]
Comment (matejonnet) [reply model/src/main/java/org/jboss/pnc/model/BuildRecord.java]: The value!

**Pull request merged**